README.md

[![Arduino CI](https://github.com/RobTillaart/M62429/workflows/Arduino%20CI/badge.svg)](https://github.com/marketplace/actions/arduino_ci)
[![Arduino-lint](https://github.com/RobTillaart/M62429/actions/workflows/arduino-lint.yml/badge.svg)](https://github.com/RobTillaart/M62429/actions/workflows/arduino-lint.yml)
[![JSON check](https://github.com/RobTillaart/M62429/actions/workflows/jsoncheck.yml/badge.svg)](https://github.com/RobTillaart/M62429/actions/workflows/jsoncheck.yml)
[![License: MIT](https://img.shields.io/badge/license-MIT-green.svg)](https://github.com/RobTillaart/M62429/blob/master/LICENSE)
[![GitHub release](https://img.shields.io/github/release/RobTillaart/M62429.svg?maxAge=3600)](https://github.com/RobTillaart/M62429/releases)


# M62429

Arduino library for M62429 volume control IC.


## Description

This library is used to set the attenuation (volume) of the 
M62429 IC a.k.a. FM62429.

The communication needs a minimum delay of 1.6 microseconds. 
This is defined in the library in **M62429_CLOCK_DELAY** == 2
For AVR (UNO, slow device) it is defined as 0, as the digitalWrite
takes time enough. 

For faster processors this define can be overruled runtime by setting it 
before including "M62429.h" or by defining it as command line parameter.


## Interface

The interface is straightforward

- **void begin(uint8_t dataPin, uint8_t clockPin)** defines the clock and data pin.
One has to create one object per IC. 
- **int getVolume(uint8_t channel)** channel is 0 or 1 or 2 (both). 
In the latter case the volume of channel 0 is used as volume of both channels.
- **int setVolume(uint8_t channel, uint8_t volume)** 
  - channel = { 0, 1, 2 = both; volume = {0 .. 255 }
  - Note: if system is muted, no changes are made.
- **int incr()** increment volume of both channels until max is reached.
This is another way to set volume that is better suited for a rotary 
encoder or a \[+\] button
  - Note: if system is muted, no changes are made.
- **int decr()** decrement volume of both channels until 0 is reached. See **incr()**.
  - Note: if system is muted, no changes are made.
- **int average()** averages the 2 channels to same = average level.  
Sort of set balance in the middle functionality.
  - Note: if system is muted, no changes are made.
- **void muteOn()** silences both channels but remembers the volume..
GetVolume() will return the 'saved' volume value.
- **void muteOff()** resets the volume per channel again.
- **bool isMuted()** returns the muted state. 

Note: the volume goes from 0..255 while the actual steps go from 0..87.
Therefore not every step in volume will make a "real" step (roughly 1 in 3).


#### Error codes

The functions **getVolume(), setVolume(), incr(), decr()** and **average()**
can return one of the  error codes.


| value | name                 | notes     |
|:-----:|:---------------------|:----------|
|   0   | M62429_OK            | no error  |
|  -1   | M62429_MUTED         | system is muted, use **muteOff()** |
|  -10  | M62429_CHANNEL_ERROR | channel must be 0, 1 or 2          |


## Operation

See examples


## Future

#### Mixer

- Control multiple M62429 IC's 
- One shared dataPin and clockPin per IC. 
- use a PCF8574 / PCF8575 as selector 
- would allow for 16 stereo or 32 mono channels. 
Runtime configuration mono / stereo would be cool.

```
  PCF8574 or PCF8575 is used as device selector
  AND port takes DATA and CLOCK to right M62429
  identical schema for all 8/16 ports

             PCF8575                 AND             M62429
           +---------+              +---+          +---------+
           |         |=======+======| & |==========| data    |
           |         |   +===|======|   |          |         |
   - I2C - |         |   |   |      +---+          |         |
           |         |   |   |                     |         |
           |         |   |   |      +---+          |         |
           |         |   |   +======| & |==========| clock   |
           |         |   |     +====|   |          |         |
           |         |   |     |    +---+          |         |
           +---------+   |     |                   +---------+
                         |     |
     DATA  ==============+     |
     CLOCK ====================+
```


#### Volume pan model

- model with one **volume(0..100%)** and one **balance(-100..100)** == **pan()**.  
Also a **left()** and **right()** incremental balance might be added.
This could work better than 2 separate volume channels.


#### Other

- change **getVolume(both)** to return max of the two channels?
  would be better.
- **min()** to reduce volume to minimum of both channels.
- **incr()** and **decr()** could have channel parameter.
default is both (backwards compatible)


**wont**
- **muteOff()** should increase gradually.  takes too much blocking time.
- **Mute()** could be per channel, default = both / all.
would add a lot of extra testing. the user can implement a **setVOlume(chan, 0)**
- **mute50()** reduces levels with 50% (rounded down?).
  user can implement this **setVolume(chan, getVolume(chan)/2);**



CHANGELOG.md
# Change Log M62529

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).


## [0.3.6] - 2022-11-15
- Add RP2040 support to build-CI.
- Add CHANGELOG.md


## [0.3.5] - 2022-02-21
- add minimal M62429_RAW class

## [0.3.4] - 2022-02-15
- improve conditional delayMicroseconds()

## [0.3.3] - 2022-02-12  
- fix #10 add channel parameter to incr() and decr()
- improve muteOn() + muteOff()

## [0.3.2] - 2022-02-12
- fix #8 mute behaviour

## [0.3.1] - 2022-02-12
- performance updates
- average(), incr() and decr()

## [0.3.0] - 2021-12-30  
- fix #4 incomplete protocol
- add examples
- refactor

----

## [0.2.3] - 2021-12-21
- update library.json, license, readme, minor edits

## [0.2.2] - 2021-05-27
- fix library.properties

## [0.2.1] - 2020-12-30
- add Arduino-CI + unit test

## [0.2.0] - 2020-08-02
- refactor

----

## [0.1.0] - 2019-01-17
- initial version




test/unit_test_001.cpp
//
//    FILE: unit_test_001.cpp
//  AUTHOR: Rob Tillaart
//    DATE: 2020-12-30
// PURPOSE: unit tests for the M62429 volume IC
//          https://github.com/RobTillaart/M62429
//          https://github.com/Arduino-CI/arduino_ci/blob/master/REFERENCE.md
//

// supported assertions
// ----------------------------
// assertEqual(expected, actual);               // a == b
// assertNotEqual(unwanted, actual);            // a != b
// assertComparativeEquivalent(expected, actual);    // abs(a - b) == 0 or (!(a > b) && !(a < b))
// assertComparativeNotEquivalent(unwanted, actual); // abs(a - b) > 0  or ((a > b) || (a < b))
// assertLess(upperBound, actual);              // a < b
// assertMore(lowerBound, actual);              // a > b
// assertLessOrEqual(upperBound, actual);       // a <= b
// assertMoreOrEqual(lowerBound, actual);       // a >= b
// assertTrue(actual);
// assertFalse(actual);
// assertNull(actual);

// // special cases for floats
// assertEqualFloat(expected, actual, epsilon);    // fabs(a - b) <= epsilon
// assertNotEqualFloat(unwanted, actual, epsilon); // fabs(a - b) >= epsilon
// assertInfinity(actual);                         // isinf(a)
// assertNotInfinity(actual);                      // !isinf(a)
// assertNAN(arg);                                 // isnan(a)
// assertNotNAN(arg);                              // !isnan(a)

#include <ArduinoUnitTests.h>


#include "Arduino.h"
#include "M62429.h"



unittest_setup()
{
  fprintf(stderr, "M62429_VERSION: %s\n", (char *) M62429_VERSION);
}


unittest_teardown()
{
}


/*
unittest(test_new_operator)
{
  assertEqualINF(exp(800));
  assertEqualINF(0.0/0.0);
  assertEqualINF(42);

  assertEqualNAN(INFINITY - INFINITY);
  assertEqualNAN(0.0/0.0);
  assertEqualNAN(42);
}
*/


unittest(test_setVolume)
{
  uint8_t dataPin = 6;
  uint8_t clockPin = 7;

  M62429 AMP;
  AMP.begin(dataPin, clockPin);

  fprintf(stderr, "\nset get volume(0)\n");
  for (int vol = 0; vol < 10; vol++)
  {
    AMP.setVolume(0, vol);
    assertEqual(vol, AMP.getVolume(0));
  }

  fprintf(stderr, "\nset get volume(1)\n");
  for (int vol = 10; vol < 20; vol++)
  {
    AMP.setVolume(1, vol);
    assertEqual(vol, AMP.getVolume(1));
  }

  fprintf(stderr, "\nset get volume(2)\n");
  for (int vol = 30; vol < 40; vol++)
  {
    AMP.setVolume(2, vol);
    assertEqual(vol, AMP.getVolume(0));
    assertEqual(vol, AMP.getVolume(1));
  }

  fprintf(stderr, "\nerror volume(3)\n");
  assertEqual(M62429_CHANNEL_ERROR, AMP.setVolume(3, 25));
}


unittest(test_incr_decr)
{
  uint8_t dataPin = 6;
  uint8_t clockPin = 7;

  M62429 AMP;
  AMP.begin(dataPin, clockPin);

  fprintf(stderr, "\nincr decr channel 0\n");
  AMP.setVolume(2, 0);
  for (int i = 0; i < 10; i++) AMP.incr(0);
  assertEqual(10, AMP.getVolume(0));
  assertEqual(0,  AMP.getVolume(1));
  for (int i = 0; i < 5; i++) AMP.decr(0);
  assertEqual(5, AMP.getVolume(0));
  assertEqual(0, AMP.getVolume(1));

  fprintf(stderr, "\nincr decr channel 1\n");
  AMP.setVolume(2, 0);
  for (int i = 0; i < 10; i++) AMP.incr(1);
  assertEqual(0,  AMP.getVolume(0));
  assertEqual(10, AMP.getVolume(1));
  for (int i = 0; i < 5; i++) AMP.decr(1);
  assertEqual(5, AMP.getVolume(0));
  assertEqual(5, AMP.getVolume(1));

  fprintf(stderr, "\nincr decr channel 2\n");
  AMP.setVolume(2, 0);
  for (int i = 0; i < 10; i++) AMP.incr();  // 2 is default
  assertEqual(10, AMP.getVolume(0));
  assertEqual(10, AMP.getVolume(1));
  for (int i = 0; i < 5; i++) AMP.decr();
  assertEqual(5, AMP.getVolume(0));
  assertEqual(5, AMP.getVolume(1));

  fprintf(stderr, "\nincr decr channel error\n");
  assertEqual(M62429_CHANNEL_ERROR, AMP.incr(3));
  assertEqual(M62429_CHANNEL_ERROR, AMP.decr(3));
}


unittest(test_average)
{
  uint8_t dataPin = 6;
  uint8_t clockPin = 7;

  M62429 AMP;
  AMP.begin(dataPin, clockPin);

  AMP.setVolume(0, 10);
  AMP.setVolume(1, 20);
  assertEqual(10, AMP.getVolume(0));
  assertEqual(20, AMP.getVolume(1));
  AMP.average();
  assertEqual(15, AMP.getVolume(0));
  assertEqual(15, AMP.getVolume(1));
}


unittest(test_mute)
{
  uint8_t dataPin = 6;
  uint8_t clockPin = 7;

  M62429 AMP;
  AMP.begin(dataPin, clockPin);

  fprintf(stderr, "mute on off\n");
  assertFalse(AMP.isMuted());
  AMP.muteOn();
  assertTrue(AMP.isMuted());
  AMP.muteOff();
  assertFalse(AMP.isMuted());

  assertEqual(M62429_OK, AMP.setVolume(0, 10));
  assertEqual(10, AMP.getVolume(0));

  fprintf(stderr, "mute on\n");
  AMP.muteOn();
  assertTrue(AMP.isMuted());

  assertEqual(M62429_MUTED, AMP.setVolume(0, 20));
  assertEqual(10, AMP.getVolume(0));

  assertEqual(M62429_MUTED, AMP.setVolume(0, 0));
  assertEqual(10, AMP.getVolume(0));

  assertEqual(M62429_MUTED, AMP.incr());
  assertEqual(10, AMP.getVolume(0));

  assertEqual(M62429_MUTED, AMP.decr());
  assertEqual(10, AMP.getVolume(0));

  assertEqual(M62429_MUTED, AMP.average());
  assertEqual(10, AMP.getVolume(0));

  AMP.muteOff();
  assertFalse(AMP.isMuted());
  assertEqual(M62429_OK, AMP.setVolume(0, 20));
  assertEqual(20, AMP.getVolume(0));

}


unittest_main()

// --------


M62429.h
#pragma once
//
//    FILE: M62429.h
//  AUTHOR: Rob Tillaart
// PURPOSE: Arduino library for M62429 volume control IC
// VERSION: 0.3.0
//
// HISTORY: See M62429.cpp
//     URL: https://github.com/RobTillaart/M62429


#include "Arduino.h"


#define M62429_VERSION              (F("0.3.0"))


// minimum pulse width CLOCK = 1.6 us (datasheet);
// digitalWrite takes enough time on UNO / AVR so clock_delay == 0
// Note that if clock pulses are long enough the data pulses are too.
#ifndef M62429_CLOCK_DELAY
#ifdef __AVR__
#define M62429_CLOCK_DELAY          0
#else
#define M62429_CLOCK_DELAY          2
#endif
#endif


// ERROR CODES
#define M62429_MUTED                -1
#define M62429_CHANNEL_ERROR        -10


class M62429
{
public:

  void begin(uint8_t dataPin, uint8_t clockPin);

  // channel = { 0, 1 }
  int     getVolume(uint8_t channel);
  // channel = { 0, 1, 2 = both; volume = {0 .. 255 }
  int     setVolume(uint8_t channel, uint8_t volume);
  // increment /decrement until max/min
  void    incr();
  void    decr();
  void    average();

  void    muteOn();
  void    muteOff();
  bool    isMuted()  { return _muted; };

private:
  uint8_t _vol[2] = { 0, 0 };
  uint8_t _data   = 0;
  uint8_t _clock  = 0;
  bool    _muted  = false;

  void    _setVolume(uint8_t channel, uint8_t volume);
};


// -- END OF FILE --



M62429.cpp
//
//    FILE: M62429.cpp
//  AUTHOR: Rob Tillaart
// PURPOSE: Arduino library for M62429 volume control IC
// VERSION: 0.3.3
// HISTORY: See M62429.cpp2
//     URL: https://github.com/RobTillaart/M62429

//  HISTORY:
//  0.1.0   2019-01-17  initial version
//  0.2.0   2020-08-02  refactor
//  0.2.1   2020-12-30  add Arduino-CI + unit test
//  0.2.2   2021-05-27  fix library.properties
//  0.2.3   2021-12-21  update library.json, license, readme, minor edits
//  0.3.0   2021-12-30  fix #4 incomplete protocol
//                      add examples, refactor
//  0.3.1   2022-02-12  performance updates
//                      average(), incr() and decr()
//  0.3.2   2022-02-12  fix #8 mute behaviour
//  0.3.3   2022-02-12  fix #10 add channel parameter to incr() and decr()
//                      improve muteOn() + muteOff()


#include "M62429.h"

#define M62429_MAX_ATTN           87 // decibel


void M62429::begin(uint8_t dataPin, uint8_t clockPin)
{
  _data = dataPin;
  _clock = clockPin;
  pinMode(_data, OUTPUT);
  pinMode(_clock, OUTPUT);
  digitalWrite(_data, LOW);
  digitalWrite(_clock, LOW);

  _muted = false;
  setVolume(2, 0);
}


int M62429::getVolume(uint8_t channel)
{
  if (channel > 2) return M62429_CHANNEL_ERROR;
  return _vol[channel & 1];
}


int M62429::setVolume(uint8_t channel, uint8_t volume)
{
  if (channel > 2) return M62429_CHANNEL_ERROR;
  if (_muted)      return M62429_MUTED;

  uint16_t attn = (M62429_MAX_ATTN * volume)/255;
  _setAttn(channel, attn);

  // update cached values
  if (channel == 0)      _vol[0] = volume;
  else if (channel == 1) _vol[1] = volume;
  else                   _vol[0] = _vol[1] = volume;
  return M62429_OK;
}


int M62429::incr(uint8_t channel)
{
  if (channel > 2) return M62429_CHANNEL_ERROR;
  if (_muted) return M62429_MUTED;

  if ( ((channel == 0) || (channel == 2))  && (_vol[0] < 255))
  {
    _vol[0]++;
    setVolume(0, _vol[0]);
  }
  if ( ((channel == 1) || (channel == 2)) && (_vol[1] < 255))
  {
    _vol[1]++;
    setVolume(1, _vol[1]);
  }
  return M62429_OK;
}


int M62429::decr(uint8_t channel)
{
  if (channel > 2) return M62429_CHANNEL_ERROR;
  if (_muted) return M62429_MUTED;

  if ( ((channel == 0) || (channel == 2)) && (_vol[0] > 0))
  {
    _vol[0]--;
    setVolume(0, _vol[0]);
  }
  if ( ((channel == 1) || (channel == 2)) && (_vol[1] > 0))
  {
    _vol[1]--;
    setVolume(1, _vol[1]);
  }
  return M62429_OK;
}


int M62429::average()
{
  if (_muted) return M62429_MUTED;
  uint8_t v = (((int)_vol[0]) + _vol[1]) / 2;
  setVolume(2, v);
  return M62429_OK;
}


void M62429::muteOn()
{
  if (_muted) return;
  _muted = true;
  if ((_vol[0] > 0) || (_vol[1] > 0)) _setAttn(2, 0);
}


void M62429::muteOff()
{
  if (_muted == false) return;
  _muted = false;
  if (_vol[0] > 0) setVolume(0, _vol[0]);
  if (_vol[1] > 0) setVolume(1, _vol[1]);
}


////////////////////////////////////////////////////////////////////
//
// PRIVATE
//

// attn = 0..M62429_MAX_ATTN
void M62429::_setAttn(uint8_t channel, uint8_t attn)
{
  uint16_t databits = 0x0200;               //         D9 latch bit
  databits |= ((attn & 0x03) << 7);         //  D8  -  D7
  databits |= (attn & 0x7C);                //  D6  -  D2
  // channel == 2 -> both 0x00 is default
  if (channel == 0) databits |= 0x03;       //  D0  -  D1
  if (channel == 1) databits |= 0x02;       //  D0  -  D1

  // write D0 - D9
  for (uint8_t i = 0; i < 10; i++)
  {
    digitalWrite(_data, databits & 0x01);
    databits >>= 1;
    digitalWrite(_clock, HIGH);
    // Note if _clock pulses are long enough, _data pulses are too.
    if (M62429_CLOCK_DELAY > 0) delayMicroseconds(M62429_CLOCK_DELAY);

    digitalWrite(_data, LOW);
    digitalWrite(_clock, LOW);
    if (M62429_CLOCK_DELAY > 0) delayMicroseconds(M62429_CLOCK_DELAY);
  }

  // Send D10 HIGH bit (Latch signal)
  digitalWrite(_data, HIGH);
  digitalWrite(_clock, HIGH);
  if (M62429_CLOCK_DELAY > 0) delayMicroseconds(M62429_CLOCK_DELAY);

  // latch D10  signal requires _clock low before _data
  // make _data dummy write to keep timing constant
  digitalWrite(_data, HIGH);
  digitalWrite(_clock, LOW);
  if (M62429_CLOCK_DELAY > 0) delayMicroseconds(M62429_CLOCK_DELAY);

  digitalWrite(_data, LOW);
  if (M62429_CLOCK_DELAY > 0) delayMicroseconds(M62429_CLOCK_DELAY);
}


// -- END OF FILE --

